src/Firestore/QuerySnapshot.php

```diff
@@ -0,0 +1,240 @@
+<?php
+/**
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Firestore;
+
+use Grpc;
+use Google\Cloud\Core\ExponentialBackoff;
+use Google\Cloud\Core\Exception\ServiceException;
+use Google\Cloud\Firestore\Connection\ConnectionInterface;
+
+/**
+ * Represents the result set of a Cloud Firestore Query.
+ *
+ * Example:
+ * ```
+ * use Google\Cloud\Firestore\FirestoreClient;
+ *
+ * $firestore = new FirestoreClient();
+ * $collection = $firestore->collection('users');
+ * $query = $collection->where('age', '>', 18);
+ *
+ * $snapshot = $query->documents();
+ * ```
+ *
+ * ```
+ * // Snapshots can be iterated with foreach:
+ * foreach ($snapshot as $user) {
+ *     echo $user['name'] . PHP_EOL;
+ * }
+ * ```
+ */
+class QuerySnapshot implements \IteratorAggregate
```

Comment:
Note that this is very different from how the other clients expose Query results.

Can you check the newly added section on the Watch API in the Server API doc and confirm that your design will work as is? Watch adds a 'changes' list to this class and needs to provide its own sorting for the 'docs' array.

If that can be adopted to work without breaking the current design, then I would think that it is ok to keep as is.

Reply:
The original implementation was attempting to make use of the server stream, similarly to how we implement query results in Spanner. I've done away with that, as without resume tokens and the like, a results stream doesn't make a ton of sense, and it brought in some weird stuff to get retries right if the stream broke before being finished. I think you'll find the new implementation is closer to the other libraries.